Epoch 2.4: Re-enabling PoX via PoX-3

kantai opened the pull request:
This PR implements Epoch 2.4 as proposed in SIP-22 and updated in SIP-24.

Commit:
introduce epoch 2.4

Commit:
add STXBalance::LockedPoxThree variant

pavitthrap committed:
wip - add pox 3 contract

Commit:
add comments to epoch_22 test

pavitthrap committed:
added burnstateDB method to create getter for pox_3 contract activation

pavitthrap committed:
fix to test

pavitthrap committed:
applying cargo fmt

obycode committed:
fix: add `delegated-to` field to `stacking-state`

This is part of SIP-022 to fix the problems with `stack-increase`. This
field allows us to track whether the stacker is stacking directly or
through a pool, and ensure that they cannot stack more than one way at a
time. More testing is still needed.

pavitthrap committed:
adjust first pox 3 cycle

pavitthrap committed:
added special handlers

pavitthrap committed:
used constant

Commit:
address PR nits

obycode committed:
fix: fix error in `increase-reward-cycle-entry`

This fixes the bug discussed in SIP-022, where the total stacked ustx
amount was applied to the stacker, instead of just that stacker's
amount.

Commit:
introduce epoch 2.4

pavitthrap committed:
made changes needed for epoch 2.4

Commit:
fix test build errors, add receipt

obycode committed:
feat(pox-3): add more assertions

obycode committed:
refactor(pox-3): `first-extend-cycle`

Simplify the usage of this variable due to the simplifications available
for pox-3.

zone117x committed:
feat: add `pox-3` to contract list in `/v2/pox` RPC response

obycode committed:
chore: add comment on `first-extend-cycle`

pavitthrap committed:
respond to comments

pavitthrap committed:
fixed test

zone117x committed:
feat: allow epoch2.2 through epoch2.4 to be configured in regtest mode (AKA neon/krypton)

zone117x committed:
fix: v2_unlock_height should be Epoch2.2 start_height+1

Commit:
expanded epoch validation checks

zone117x committed:
chore: PR feedback (pox_3_first_cycle off-by-one)

pavitthrap committed:
removed burnchain height check

zone117x committed:
chore: fix debug message for v2_unlock_height override

zone117x committed:
Revert "chore: PR feedback (pox_3_first_cycle off-by-one)"

This reverts commit 5e6c36ed8076ae5ecf071e2c9a087d18fe8ea517.

zone117x committed:
chore: PR feedback (pox_3_first_cycle off-by-one)

Commit:
use pox-3 for reward sets after activation height

Commit:
initial testing

Commit:
working invariants in the simple test case

kantai commented:
Heads up @obycode -- I pushed a few changes to pox-3.clar in db2e35f, it was failing with some simple analysis errors.

friedger commented on src/chainstate/stacks/boot/pox-3.clar:
```diff
@@ -0,0 +1,1316 @@
+
+      ;; standard can-stack-stx checks
+      (try! (can-stack-stx pox-addr amount-ustx first-extend-cycle lock-period))
+
```
Does it matter if the stacker changes pox-address through extending?

kantai replied:
I don't think it matters, but it's probably wise to err on the side of safety and enforce using the same pox-address.

friedger commented on src/chainstate/stacks/boot/pox-3.clar:
```diff
@@ -0,0 +1,1316 @@
+                            (>= until-burn-ht unlock-height)
+                        true)
+                  (err ERR_DELEGATION_EXPIRES_DURING_LOCK)))
+
```
Assert that stacker-info delegate-to is equal to the tx-sender to avoid swimming in two pools and risking the auto-unlock bug:
Alice delegates 100 STX to Bob
Bob locks Alice 100 STX
Alice revokes and delegates 150 STX to Cloe
Cloe increase Alice stacking by 50.
If Bob's pool misses the minimum, then Alice earns rewards for 150 STX and has only 50 STX locked.

kantai replied:
I think that assertion is performed at lines 1132-1136 in this method.

friedger commented on src/chainstate/stacks/boot/pox-3.clar:
```diff
@@ -0,0 +1,1316 @@
+      (asserts! (is-eq (len (get reward-set-indexes stacker-state)) u0)
+                (err ERR_STACKING_NOT_DELEGATED))
+
+      ;; stacker must be delegated to tx-sender
```
;; stacker STX must have been locked initially by tx-sender

This implies that changing the pool requires 1 cool down cycle! This should go somewhere into the change log of this contract.

friedger replied:
This is a problem for sBTC mini. The current design requires that users can switch pools each cycle.

friedger commented on src/chainstate/stacks/boot/pox-3.clar:
```diff
@@ -0,0 +1,1316 @@
+;; The .pox-3 contract
```
Add changelog from pox and pox-2

Commit:
implement pox-3 auto-unlock, add pox-3 auto unlock test

pavitthrap committed:
added epoch2.4 test

disable_pox_v2 is failing only on final assertion, not crashing

wip, trying to get epoch 2.4 to start right

rename test, revert epoch_22

fix_to_pox_contract passes

auto unlock test passing

added more asserts for auto unlock test

jcnelson commented:
That's going to be a problem for sBTC mini. sBTC mini expects
delegate-stack-extend to be able to change the PoX address for the next
cycle while the STX are stacked for the current cycle.

On Thu, May 4, 2023, 4:59 PM Aaron Blankstein ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/chainstate/stacks/boot/pox-3.clar
> <https://github.com/stacks-network/stacks-blockchain/pull/3699#discussion_r1185508257>
> :
>
> > +
> +      ;; must be called directly by the tx-sender or by an allowed contract-caller
> +      (asserts! (check-caller-allowed)
> +                (err ERR_STACKING_PERMISSION_DENIED))
> +
> +      ;; tx-sender must be locked
> +      (asserts! (> amount-ustx u0)
> +        (err ERR_STACK_EXTEND_NOT_LOCKED))
> +
> +      ;; tx-sender must not be delegating
> +      (asserts! (is-none (get-check-delegation tx-sender))
> +        (err ERR_STACKING_ALREADY_DELEGATED))
> +
> +      ;; standard can-stack-stx checks
> +      (try! (can-stack-stx pox-addr amount-ustx first-extend-cycle lock-period))
> +
>
> I don't think it matters, but it's probably wise to err on the side of
> safety and enforce using the same pox-address.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stacks-network/stacks-blockchain/pull/3699#discussion_r1185508257>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQJK7EZMJMP3LNFCUOLFTXEQKEJANCNFSM6AAAAAAXTRKVYI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>

kantai commented:
> That's going to be a problem for sBTC mini. sBTC mini expects delegate-stack-extend to be able to change the PoX address for the next cycle while the STX are stacked for the current cycle.

Doesn't sBTC mini also require the ability to slash lockups? I think that would require other changes as well.

Commit:
fix pox-2 tests for the updated invariant checks

jcnelson commented:
No, sBTC mini merely sets the PoX payout address to the burn address on
protocol failure. It only affects stackers who have delegated to the sBTC
contract.

On Thu, May 4, 2023, 6:29 PM Aaron Blankstein ***@***.***>
wrote:

> That's going to be a problem for sBTC mini. sBTC mini expects
> delegate-stack-extend to be able to change the PoX address for the next
> cycle while the STX are stacked for the current cycle.
>
> Doesn't sBTC mini also require the ability to slash lockups? I think that
> would require other changes as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stacks-network/stacks-blockchain/pull/3699#issuecomment-1535488640>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQJK6VRDWE575TWT5X433XEQUWDANCNFSM6AAAAAAXTRKVYI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

igorsyl commented:
I organized a meeting this morning with @friedger, @setzeus, @jo-tm, @radicleart, and @netrome we agreed that we will not deploy sBTC Mini to Stacks primary testnet. Instead, we will create a new Stacks testnet network specifically for sBTC, where we can develop the POX-4 contract that will include fixes to the design issues outlined here.

This decision will decouple the design of POX-3 being discussed here from sBTC's contract design so you can fully focus on reactivating POX via POX-3 and not worry about the impact on sBTC Mini.

[Meeting notes](https://docs.google.com/document/d/1x4djVyG5ymQ5F8-wnlUzSz1jT9hdu6AZdyPxZeGmFHI/edit)

cc @diwakergupta @muneeb-ali @andresgalante

jcnelson commented:
What's the point of sBTC mini then? If you're going to have a dedicated
testnet, it might as well be one for the real sBTC.  sBTC mini was designed
to allow it to be deployed on the mainline testnet (or even mainnet)
without consensus changes.

On Fri, May 5, 2023, 8:49 AM Igor Sylvester ***@***.***>
wrote:

> I organized a meeting this morning with @friedger
> <https://github.com/friedger>, @setzeus <https://github.com/setzeus>,
> @jo-tm <https://github.com/jo-tm>, @radicleart
> <https://github.com/radicleart>, and @netrome <https://github.com/netrome>
> we agreed that we will not deploy sBTC Mini to Stacks primary testnet.
> Instead, we will create a new Stacks testnet network specifically for sBTC,
> where we can develop the POX-4 contract that will include fixes to the
> design issues outlined here.
>
> This decision will decouple the design of POX-3 being discussed here from
> sBTC's contract design so you can fully focus on reactivating POX via POX-3
> and not worry about the impact on sBTC Mini.
>
> Meeting notes
> <https://docs.google.com/document/d/1x4djVyG5ymQ5F8-wnlUzSz1jT9hdu6AZdyPxZeGmFHI/edit>
>
> cc @diwakergupta <https://github.com/diwakergupta> @muneeb-ali
> <https://github.com/muneeb-ali> @andresgalante
> <https://github.com/andresgalante>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stacks-network/stacks-blockchain/pull/3699#issuecomment-1536214479>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQJK4QXUC32KJVHS4E4OLXETZOPANCNFSM6AAAAAAXTRKVYI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Commit:
fix bug in pox-3-increase handler, add test for delegate-stack-increase

Commit:
test: add pox_3_tests::stack_increase

igorsyl commented:
> What's the point of sBTC mini then? If you're going to have a dedicated testnet, it might as well be one for the real sBTC. sBTC mini was designed to allow it to be deployed on the mainline testnet (or even mainnet) without consensus changes.
> […](#)
> On Fri, May 5, 2023, 8:49 AM Igor Sylvester ***@***.***> wrote: I organized a meeting this morning with @friedger <https://github.com/friedger>, @setzeus <https://github.com/setzeus>, @jo-tm <https://github.com/jo-tm>, @radicleart <https://github.com/radicleart>, and @netrome <https://github.com/netrome> we agreed that we will not deploy sBTC Mini to Stacks primary testnet. Instead, we will create a new Stacks testnet network specifically for sBTC, where we can develop the POX-4 contract that will include fixes to the design issues outlined here. This decision will decouple the design of POX-3 being discussed here from sBTC's contract design so you can fully focus on reactivating POX via POX-3 and not worry about the impact on sBTC Mini. Meeting notes <https://docs.google.com/document/d/1x4djVyG5ymQ5F8-wnlUzSz1jT9hdu6AZdyPxZeGmFHI/edit> cc @diwakergupta <https://github.com/diwakergupta> @muneeb-ali <https://github.com/muneeb-ali> @andresgalante <https://github.com/andresgalante> — Reply to this email directly, view it on GitHub <[#3699 (comment)](https://github.com/stacks-network/stacks-blockchain/pull/3699#issuecomment-1536214479)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADQJK4QXUC32KJVHS4E4OLXETZOPANCNFSM6AAAAAAXTRKVYI> . You are receiving this because you commented.Message ID: ***@***.***>

Replied via Discord: https://discord.com/channels/621759717756370964/1100888369380212757/1104037347932250122

jcnelson commented:
Please have the courtesy to copy the text to GitHub so those who are not
already in that channel know which messages are relevant.

On Fri, May 5, 2023, 12:25 PM Igor Sylvester ***@***.***>
wrote:

> What's the point of sBTC mini then? If you're going to have a dedicated
> testnet, it might as well be one for the real sBTC. sBTC mini was designed
> to allow it to be deployed on the mainline testnet (or even mainnet)
> without consensus changes.
> … <#m_-3257510939477943128_>
> On Fri, May 5, 2023, 8:49 AM Igor Sylvester *@*.*> wrote: I organized a
> meeting this morning with @friedger <https://github.com/friedger>
> https://github.com/friedger <https://github.com/friedger>, @setzeus
> <https://github.com/setzeus> https://github.com/setzeus
> <https://github.com/setzeus>, @jo-tm <https://github.com/jo-tm>
> https://github.com/jo-tm <https://github.com/jo-tm>, @radicleart
> <https://github.com/radicleart> https://github.com/radicleart
> <https://github.com/radicleart>, and @netrome <https://github.com/netrome>
> https://github.com/netrome <https://github.com/netrome> we agreed that we
> will not deploy sBTC Mini to Stacks primary testnet. Instead, we will
> create a new Stacks testnet network specifically for sBTC, where we can
> develop the POX-4 contract that will include fixes to the design issues
> outlined here. This decision will decouple the design of POX-3 being
> discussed here from sBTC's contract design so you can fully focus on
> reactivating POX via POX-3 and not worry about the impact on sBTC Mini.
> Meeting notes
> https://docs.google.com/document/d/1x4djVyG5ymQ5F8-wnlUzSz1jT9hdu6AZdyPxZeGmFHI/edit
> <https://docs.google.com/document/d/1x4djVyG5ymQ5F8-wnlUzSz1jT9hdu6AZdyPxZeGmFHI/edit>
> cc @diwakergupta <https://github.com/diwakergupta>
> https://github.com/diwakergupta <https://github.com/diwakergupta>
> @muneeb-ali <https://github.com/muneeb-ali> https://github.com/muneeb-ali
> <https://github.com/muneeb-ali> @andresgalante
> <https://github.com/andresgalante> https://github.com/andresgalante
> <https://github.com/andresgalante> — Reply to this email directly, view it
> on GitHub <#3699 (comment)
> <https://github.com/stacks-network/stacks-blockchain/pull/3699#issuecomment-1536214479>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AADQJK4QXUC32KJVHS4E4OLXETZOPANCNFSM6AAAAAAXTRKVYI
> <https://github.com/notifications/unsubscribe-auth/AADQJK4QXUC32KJVHS4E4OLXETZOPANCNFSM6AAAAAAXTRKVYI>
> . You are receiving this because you commented.Message ID: @.*>
>
> Replied via Discord:
> https://discord.com/channels/621759717756370964/1100888369380212757/1104037347932250122
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stacks-network/stacks-blockchain/pull/3699#issuecomment-1536492791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQJK3REVMCFZU35MBLZUDXEUSXHANCNFSM6AAAAAAXTRKVYI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Commit:
test: more porting of pox-2 tests to pox-3

obycode committed:
chore: fix typos

Commit:
test: add segwit, pox-address get-burn-info validation tests

Commit:
tests: finish porting pox_2_tests to pox_3

Commit:
update tests with new default regtest epochs

friedger commented on src/chainstate/stacks/boot/pox_3_tests.rs:
```diff
@@ -0,0 +1,4394 @@
+        latest_block = peer.tenure_with_txs(&[], &mut coinbase_nonce);
+    }
+
+    // in the next tenure, PoX 2 should now exist.
```
PoX 3 should exist

friedger commented on src/chainstate/stacks/boot/pox_3_tests.rs:
```diff
@@ -0,0 +1,4394 @@
+    // first tenure is empty
+    let mut latest_block = peer.tenure_with_txs(&[], &mut coinbase_nonce);
+
+    // Roll to Epoch-2.4 and perform the delegate-stack-extend tests
```
// Roll to Epoch-2.4 ~~and perform the delegate-stack-extend tests~~

friedger reviewed (commented):
I read through pox_3_tests to learn and left some comments to help understanding the code.
Mainly copy paste errors.

pavitthrap committed:
fixed test, added to gh workflow

obycode committed:
chore: update testnet 2.4 activation

This puts it before the next prepare phase.

obycode committed:
Merge pull request #3716 from stacks-network/chore/2.4-testnet

Update 2.4 testnet activation height

Commit:
feat: implement proposed SIP-024 sanitization logic for epoch-2.4

friedger commented on src/chainstate/stacks/boot/pox-3.clar:
```diff
@@ -0,0 +1,1316 @@
+                (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; tx-sender principal must not be stacking
+      (asserts! (is-none (get-stacker-info tx-sender))
```
Why that? This is a regression. In pox-2 it was possible to change the delegation amount while stacking.
https://github.com/stacks-network/stacks-blockchain/blob/d657bdde9bbe263050d2e7ffa4b822fdaefa2530/src/chainstate/stacks/boot/pox-2.clar#L651

obycode replied:
This avoids the problem that was present in auto-unlock.

friedger replied:
It takes a lot of power from delegate-stack-increase

friedger replied:
@obycode would it be enough to check that the pool does not change?

kantai replied:
I don't think so -- the issue is that `auto-unlock` relies on the invariant that if a user's solo stacking amount doesn't reach the participation threshold, then their account can be unlocked. But if they've also contributed to a pool, then their account would be unlocked _while still contributing to a reward_.

obycode replied:
Friedger and I were discussing this in Discord. What he would like is for users to be able to re-delegate to the same pool without a cooldown cycle. Essentially, this would be the equivalent of `stack-increase` and/or `stack-extend` but for delegation.

kantai replied:
But isn't that use case fulfilled with `delegate-stack-increase`?

friedger replied:
Also the current code is not convert by testing using devent-js

friedger replied:
All tests should pass if the assertion is removed. The important assertions should be in delegate-stack-* functions

Commit:
fix pox_3 auto_unlock tests

Commit:
more sanitization unit cases

pavitthrap committed:
added more test coverage for epoch 2.4

friedger committed:
fix: allow change of delegation data

friedger committed:
chore: improve comments

Commit:
address PR feedback, enable sanitization in from-consensus-buff

Commit:
Merge remote-tracking branch 'origin/feat/epoch-2.4' into test/neon-epoch-2.4-v2

Commit:
Merge pull request #3721 from friedger/fix/increase-delegation-amount

Fix/increase delegation amount

Commit:
Merge remote-tracking branch 'origin/feat/epoch-2.4' into feat/epoch-2.4-sanitize

Commit:
Merge pull request #3722 from stacks-network/feat/sanitize-test-cov

add more test coverage for epoch 2.4

Commit:
address PR feedback

Commit:
fix: affirmation calc should skip any ops whose parent is <= first_block_height

Commit:
update mainnet 2.4 height with sip-024 value

Commit:
Merge pull request #3719 from stacks-network/feat/epoch-2.4-sanitize

Feat: Add SIP-024 sanitization support to Epoch 2.4

Commit:
add comment to affirmation module explaining change

pavitthrap committed:
add epoch.truncate

Commit:
sanitize-gate the depth check

pavitthrap committed:
Merge pull request #3715 from stacks-network/test/neon-epoch-2.4-v2

Neon integration-style test for epoch 2.4

obycode reviewed (approved):
🚀

pavitthrap commented on src/chainstate/coordinator/mod.rs:
```diff
@@ -259,6 +260,7 @@ pub struct OnChainRewardSetProvider();
 impl RewardSetProvider for OnChainRewardSetProvider {
     fn get_reward_set(
         &self,
+        // Todo: `current_burn_height` is a misleading name: should be the `cycle_start_burn_height`
```
nit: address later, remove this comment

pavitthrap reviewed (approved):
lgtm

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
